README.md
# HTTP/2 Server Push

[![PHPUnit Tests](https://github.com/wearerequired/h2push/actions/workflows/phpunit-tests.yml/badge.svg)](https://github.com/wearerequired/h2push/actions/workflows/phpunit-tests.yml) [![Coding Standards](https://github.com/wearerequired/h2push/actions/workflows/coding-standards.yml/badge.svg)](https://github.com/wearerequired/h2push/actions/workflows/coding-standards.yml)

Sends Link headers to bring HTTP/2 Server Push for scripts and styles to WordPress. Falls back to `<link>` element if headers are already sent.  
Provides filters to customize and extend the resources to push.

![Screenshot](https://user-images.githubusercontent.com/617637/31279476-7c3dffd6-aaa9-11e7-91d8-57ec4435d067.png)

## Installation

Install the latest version with

`composer require wearerequired/h2push`

The plugin requires at least PHP 7.4 and WordPress 5.6.

## Hooks reference

### `h2push.as_header`

By default the plugin will use the Link header if no headers are sent yet and falls back to the `<link>` element. To change this behavior you can use the `h2push.as_header` filter. Example:

```php
// Force H2 Push to always use the `<link>` element.
add_filter( 'h2push.as_header', '__return_false' );
```

This filter is also useful if the server doesn't support HTTP/2 yet and you still want to benefit from preloading.

### `h2push.push_resources`

By default the plugin collects all enqueued scripts and styles which are have been registered before or at the `wp_enqueue_scripts` hook. The `h2push.push_resources` filters allows to customize the list of resources. Example:

```php
/**
 * Add web font and hero image to the list of resources to push/preload.
 *
 * @param array $resources List of resources.
 * @return array List of resources.
 */
function my_theme_push_resources( array $resources ): array {
	$relative_template_directory_uri = wp_parse_url( get_template_directory_uri(), PHP_URL_PATH );

	// Push web font.
	$resources[] = [
		'href' => $relative_template_directory_uri . '/assets/fonts/fancy.woff2',
		'as'   => 'font',
		'type' => 'font/woff2',
		'crossorigin',
	];

	if ( is_front_page() && ! is_paged() ) {
		// Push hero image.
		$resources[] = [
			'href' => $relative_template_directory_uri . '/assets/images/hero.webp',
			'as'   => 'image',
			'type' => 'image/webp',
		];
	}

	return $resources;
}
add_filter( 'h2push.push_resources', 'my_theme_push_resources' );
```

### `h2push.is_allowed_push_host`

By default the plugin only sends push requests for local resources where the asset URL matches the home URL. To change this behavior you can use the `h2push.is_allowed_push_host` filter. Example:

```php
/**
 * Allow resources from example.org to be pushed/preloaded too.
 *
 * @param bool   $is_allowed Whether the host should be allowed. Default true for local resources.
 * @param string $host       The host name of the resource.
 * @return bool Whether the host should be allowed.
 */
function my_theme_is_allowed_push_host( $is_allowed, $host ) {
	return $is_allowed || 'example.org' === $host;
}
add_filter( 'h2push.is_allowed_push_host', 'my_theme_is_allowed_push_host' );
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.2.0] - 2022-11-08

### Changed
* Use `script.<handle>`/`style.<handle>` as key for list of resources to make filtering via `h2push.push_resources` filter easier.
* Allow to customize the value of the `rel` attribute. Defaults to `preload`.

## [2.1.0] - 2022-07-26

### Fixed
* Fix incorrect handling of output buffer for `wp_head` action when stacked buffers are used.

### Added
* New `h2push.is_allowed_push_host` filter to allow external files to be preloaded.

## [2.0.0] - 2021-04-21

### Changed
* Requires PHP 7.4.
* Requires WordPress 5.6.
* Use output buffer for `wp_head` action to be able to send headers.

### Added
* New `h2push.as_header` filter to define whether to use Link headers or the `<link>` element.
* New `h2push.push_resources` filter to customize URLs for resources to push.
* Support resources with custom `as`, `type`, `crossorigin` and `nopush` parameters.

## [1.3.0] - 2020-09-30

### Changed
* Add fallback to `<link>` element if headers are already sent.

## [1.2.0] - 2018-06-12

### Fixed
* Fix incorrect handling of dependencies with version set to `null`.
* Fix linking of dependencies with protocol-relative URLs.

## [1.1.0] - 2017-12-09

### Changed
* Add support for relative script and style paths.

## [1.0.0] - 2017-12-08

### Added
* The initial version of the plugin.

[Unreleased]: https://github.com/wearerequired/h2push/compare/2.2.0...HEAD
[2.2.0]: https://github.com/wearerequired/h2push/compare/2.1.0...2.2.0
[2.1.0]: https://github.com/wearerequired/h2push/compare/2.0.0...2.1.0
[2.0.0]: https://github.com/wearerequired/h2push/compare/1.3.0...2.0.0
[1.3.0]: https://github.com/wearerequired/h2push/compare/1.2.0...1.3.0
[1.2.0]: https://github.com/wearerequired/h2push/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/wearerequired/h2push/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/wearerequired/h2push/compare/f1bd977d83e063311d162b2415f3499b20d7296e...0.1.0


tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package H2push
 */

$_tests_dir = getenv( 'WP_TESTS_DIR' );

if ( ! $_tests_dir ) {
	$_tests_dir = rtrim( sys_get_temp_dir(), '/\\' ) . '/wordpress-tests-lib';
}

if ( ! file_exists( $_tests_dir . '/includes/functions.php' ) ) {
	echo "Could not find $_tests_dir/includes/functions.php, have you run bin/install-wp-tests.sh ?" . PHP_EOL;
	exit( 1 );
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/h2push.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/test-sample.php
<?php
/**
 * Class SampleTest
 *
 * @package H2push
 */

/**
 * Sample test case.
 */
class SampleTest extends WP_UnitTestCase {

	/**
	 * A single example test.
	 */
	public function test_sample() {
		// Replace this with some actual testing code.
		$this->assertTrue( true );
	}
}


tests/getPushResources.php
<?php
/**
 * Class GetPushResources
 *
 * @package H2push
 */

/**
 *  Test cases for \Required\H2Push\get_push_resources().
 */
class GetPushResources extends WP_UnitTestCase {

	/**
	 * Test that \Required\H2Push\get_push_resources() exists.
	 */
	public function test_get_push_resources_exists() {
		$this->assertTrue( function_exists( '\Required\H2Push\get_push_resources' ) );
	}
}


h2push.php
<?php
/**
 * Plugin Name: HTTP/2 Server Push
 * Plugin URI: https://github.com/wearerequired/h2push/
 * Description: Sends Link headers to bring HTTP/2 Server Push for scripts and styles to WordPress.
 * Version: 2.2.0
 * Requires at least: 5.6
 * Requires PHP: 7.4
 * Author: required
 * Author URI: https://required.com
 * License: GPL-2.0+
 * License URI: http://www.gnu.org/licenses/gpl-2.0.txt
 *
 * Copyright (c) 2017-2022 required (email: info@required.ch)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2 or, at
 * your discretion, any later version, as published by the Free
 * Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

namespace Required\H2Push;

/**
 * Starts an output buffer for `wp_head`.
 *
 * The output buffer is necessary so that we can call `header()` during
 * the `wp_head` action which is the only action that allows us to collect
 * all script and style resources.
 *
 * @since 2.0.0
 */
function start_output_buffer(): void {
	ob_start();
}
add_action( 'wp_head', __NAMESPACE__ . '\start_output_buffer', 0 );

/**
 * Stops the output buffer for `wp_head`.
 *
 * @since 2.0.0
 */
function stop_output_buffer(): void {
	if ( ob_get_length() ) {
		ob_end_flush();
	}
}
add_action( 'wp_head', __NAMESPACE__ . '\stop_output_buffer', PHP_INT_MAX );

/**
 * Sends preload Link headers for script and style resources.
 *
 * @since 1.0.0
 * @since 1.3.0 Falls back to <link> element if headers are already sent.
 */
function add_link_headers(): void {
	$as_header = ! headers_sent();
	/**
	 * Filters whether to use Link headers or the <link> element.
	 *
	 * @since 2.0.0
	 *
	 * @param bool $as_header Whether to use Link headers or the <link> element.
	 */
	$as_header = apply_filters( 'h2push.as_header', $as_header );

	$resources = get_push_resources();

	foreach ( $resources as $attributes ) {
		if ( $as_header ) {
			header(
				get_link_header( $attributes ),
				false
			);
		} else {
			echo get_link_tag( $attributes ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
		}
	}
}
add_action( 'wp_head', __NAMESPACE__ . '\add_link_headers', 2, 0 );

/**
 * Builds the Link header.
 *
 * @since 2.0.0
 *
 * @param array<mixed,string> $attributes Resource attributes.
 * @return string The Link header.
 */
function get_link_header( array $attributes ): string {
	$parameters = '';

	if ( ! isset( $attributes['rel'] ) ) {
		$attributes = [ 'rel' => 'preload' ] + $attributes;
	}

	foreach ( $attributes as $attr => $value ) {
		if (
			! is_scalar( $value ) ||
			( ! \in_array( $attr, [ 'rel', 'as', 'crossorigin', 'type', 'nopush' ], true ) && ! is_numeric( $attr ) )
		) {
			continue;
		}

		if ( ! \is_string( $attr ) ) {
			$parameters .= " $value;";
		} else {
			$parameters .= " $attr=\"$value\";";
		}
	}

	$parameters = trim( $parameters );
	$parameters = rtrim( $parameters, ';' );

	return sprintf(
		'Link: <%s>; %s',
		esc_url_raw( $attributes['href'] ),
		$parameters
	);
}

/**
 * Builds the <link> tag.
 *
 * @since 2.0.0
 *
 * @param array<mixed,string> $attributes Resource attributes.
 * @return string The <link> tag.
 */
function get_link_tag( array $attributes ): string {
	$html = '';

	if ( ! isset( $attributes['rel'] ) ) {
		$attributes = [ 'rel' => 'preload' ] + $attributes;
	}

	foreach ( $attributes as $attr => $value ) {
		if (
			! is_scalar( $value ) ||
			( ! \in_array( $attr, [ 'rel', 'href', 'as', 'crossorigin', 'type' ], true ) && ! is_numeric( $attr ) )
		) {
			continue;
		}

		$value = ( 'href' === $attr ) ? esc_url( $value ) : esc_attr( $value );

		if ( ! \is_string( $attr ) ) {
			$html .= " $value";
		} else {
			$html .= " $attr='$value'";
		}
	}

	$html = trim( $html );

	return "<link $html />\n";
}

/**
 * Retrieves script and style resources that can be used for HTTP/2 Server Push.
 *
 * @since 1.0.0
 *
 * @return array<string,mixed> List of script and style resources.
 */
function get_push_resources(): array {
	global $wp_version;

	$push_resources = [];
	$home_url       = untrailingslashit( home_url() );
	$home_url_host  = wp_parse_url( $home_url, PHP_URL_HOST );

	// Stylesheets.
	$wp_styles = wp_styles();
	$wp_styles->all_deps( $wp_styles->queue );
	$files = $wp_styles->to_do;

	foreach ( $files as $i => $handle ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
		$style = $wp_styles->registered[ $handle ];
		$src   = $style->src;

		if ( empty( $src ) ) {
			continue;
		}

		// Don't push files with conditional tags.
		if ( isset( $style->extra['conditional'] ) ) {
			continue;
		}

		// Prepend a protocol.
		if ( '//' === substr( $src, 0, 2 ) || ( false === strpos( $src, ':' ) && '/' !== $src[0] ) ) {
			$src = ( is_ssl() ? 'https:' : 'http:' ) . $src;
		}

		// Check if it's a local resource.
		if ( '/' !== $src[0] ) {
			$src_host        = wp_parse_url( $src, PHP_URL_HOST );
			$is_local        = $home_url_host === $src_host;
			$is_allowed_host = apply_filters( 'h2push.is_allowed_push_host', $is_local, $src_host );
			if ( ! $is_allowed_host ) {
				continue;
			}
		}

		// Append the style version.
		if ( null !== $style->ver ) {
			$src = add_query_arg(
				'ver',
				$style->ver ? $style->ver : $wp_version,
				$src
			);
		}

		$src = str_replace( $home_url, '', $src );

		$push_resources[ 'style.' . $handle ] = [
			'href' => $src,
			'as'   => 'style',
		];
	}

	// Scripts.
	$wp_scripts = wp_scripts();
	$wp_scripts->all_deps( $wp_scripts->queue );
	$files = $wp_scripts->to_do;

	foreach ( $files as $i => $handle ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
		$script = $wp_scripts->registered[ $handle ];
		$src    = $script->src;

		if ( empty( $src ) ) {
			continue;
		}

		// Don't push files with conditional tags.
		if ( isset( $script->extra['conditional'] ) ) {
			continue;
		}

		// Prepend a protocol.
		if ( '//' === substr( $src, 0, 2 ) || ( false === strpos( $src, ':' ) && '/' !== $src[0] ) ) {
			$src = ( is_ssl() ? 'https:' : 'http:' ) . $src;
		}

		// Check if it's a local resource.
		if ( '/' !== $src[0] ) {
			$src_host        = wp_parse_url( $src, PHP_URL_HOST );
			$is_local        = $home_url_host === $src_host;
			$is_allowed_host = apply_filters( 'h2push.is_allowed_push_host', $is_local, $src_host );
			if ( ! $is_allowed_host ) {
				continue;
			}
		}

		// Append the script version.
		if ( null !== $script->ver ) {
			$src = add_query_arg(
				'ver',
				$script->ver ? $script->ver : $wp_version,
				$src
			);
		}

		$src = str_replace( $home_url, '', $src );

		$push_resources[ 'script.' . $handle ] = [
			'href' => $src,
			'as'   => 'script',
		];
	}

	/**
	 * Filters URLs for resources to preload.
	 *
	 * @param array $push_resources URLs and types for resources to preload.
	 */
	$push_resources = apply_filters( 'h2push.push_resources', $push_resources );

	return $push_resources;
}
